standard/variables.md

```diff
@@ -988,3 +998,225 @@ variable_reference
+>
+>         // context of r4 is block, ref safe context of arr[v3] is caller-context
+>         ref int r4 = ref arr[v3]; 
+>     }
+> }
+> ```
+>
+> *end example.*
+<!-- markdownlint-disable MD028 -->
+
+<!-- markdownlint-enable MD028 -->
+> *Example*: For `struct` types, the implicit `this` parameter is passed as a `ref` parameter. The ref-safe-context of the fields of a `struct` type as function-member prevents returning those fields by reference return. This rule prevents the following code:
+>
+> ```csharp
+> public struct S
+> {
+>      private int n;
+>
+>      // Disallowed: returning ref of a field.
+>      public ref int GetN() => ref n;
+> }
+>
+> public ref int M()
+> {
+>     S s = new S();
+>     ref int numRef = ref s.GetN();
+>     return ref numRef; // reference to local variable 'numRef' returned
+> }
+>
+> ```
+>
+> *end example.*
+
+#### §ref-span-safety-locals Local variable ref safe context
+
+For a local variable `v`:
+
+- If `v` is a reference variable, its ref-safe-context is the same as the ref-safe-context of its initializing expression.
+- Otherwise its ref-safe-context is the context in which it was declared.
+
+#### §ref-span-safety-parameters Parameter ref safe context
+
+For a formal parameter `p`:
+
+- If `p` is a `ref`, or `in` parameter, its ref-safe-context is the caller-context. It is safe-to-ref-return. If `p` is an `in` parameter, it can't be returned as a writable `ref` but can be returned as `ref readonly`.
+- If `p` is an `out` parameter, its ref-safe-context is the caller-context. It is safe-to-ref-return.
+- Otherwise, if `p` is the `this` parameter of a struct type, its ref-safe-context is the function-member.
+- Otherwise, the parameter is a value parameter, and its ref-safe-context is the function-member. It isn't safe-to-ref-return.
+
+#### §ref-span-safety-field-reference Field ref safe context
+
+For a variable designating a reference to a field, `e.F`:
+
+- If `e` is of a reference type, its ref-safe-context is the caller-context.
+- Otherwise, if `e` is of a value type, its ref-safe-context is the same as the ref-safe-context of `e`.
+
+#### §ref-span-safety-operators Operators
+
+The conditional operator (§12.18), `c ? ref e1 : ref e2`, and reference assignment operator, `= ref e` (§12.21.1) have reference variables as operands and yield a reference variable. For those operators, the ref-safe-context of the result is the narrowest context among the ref-safe-contexts of all `ref` operands.
+
+#### §ref-span-safety-function-invocation Function invocation
+
+For a variable `c` resulting from a ref-returning function invocation, `ref e1.M(e2, ...)`, its ref-safe-context is the narrowest of the following contexts:
+
+- The caller-context.
+- The ref-safe-context of all `ref` and `out` argument expressions (excluding the receiver).
+- For each `in` parameter of the method, if there is a corresponding expression that is a variable, its ref-safe-context, otherwise the nearest enclosing context
```

Comment:
```suggestion
- For each `in` parameter of the method, if there is a corresponding expression that is a variable, its ref-safe-context, otherwise the nearest enclosing context.
```